primitives/src/merkle_tree/internal.rs

```diff
@@ -365,61 +363,85 @@ where
 {
     /// Forget a leaf from the merkle tree. Internal branch merkle node will
     /// also be forgotten if all its leafs are forgotten.
+    #[allow(clippy::type_complexity)]
     pub(crate) fn forget_internal(
-        &mut self,
+        &self,
         height: usize,
         traversal_path: &[usize],
-    ) -> LookupResult<E, MerklePath<E, I, T>, MerklePath<E, I, T>> {
+    ) -> (
+        Arc<Self>,
+        LookupResult<E, MerklePath<E, I, T>, MerklePath<E, I, T>>,
+    ) {
         match self {
-            MerkleNode::Empty => LookupResult::NotFound(vec![MerkleNode::Empty; height + 1]),
+            MerkleNode::Empty => (
+                Arc::new(self.clone()),
+                LookupResult::NotFound(vec![MerkleNode::Empty; height + 1]),
+            ),
             MerkleNode::Branch { value, children } => {
-                match children[traversal_path[height - 1]]
-                    .forget_internal(height - 1, traversal_path)
-                {
+                let mut children = children.clone();
+                let (new_child, result) = children[traversal_path[height - 1]]
+                    .forget_internal(height - 1, traversal_path);
+                match result {
                     LookupResult::Ok(elem, mut proof) => {
                         proof.push(MerkleNode::Branch {
                             value: T::default(),
                             children: children
                                 .iter()
                                 .map(|child| {
                                     if let MerkleNode::Empty = **child {
-                                        Box::new(MerkleNode::Empty)
+                                        Arc::new(MerkleNode::Empty)
                                     } else {
-                                        Box::new(MerkleNode::ForgettenSubtree {
+                                        Arc::new(MerkleNode::ForgettenSubtree {
                                             value: child.value(),
                                         })
                                     }
                                 })
                                 .collect::<Vec<_>>(),
                         });
+                        children[traversal_path[height - 1]] = new_child;
                         if children.iter().all(|child| {
                             matches!(
                                 **child,
-                                MerkleNode::Empty | MerkleNode::ForgettenSubtree { value: _ }
+                                MerkleNode::Empty | MerkleNode::ForgettenSubtree { .. }
```

Comment:
Am I reading this right, if after doing an update all the children are empty _or_ forgotten, we replace the whole branch with a forgotten subtree? Doesn't this break non-membership proofs for the universal merkle tree? Sometimes we care to distinguish empty branches from forgotten subtrees

Reply:
Oh you just find another long standing bug.
Fixed here [5a3229c](https://github.com/EspressoSystems/jellyfish/pull/489/commits/5a3229c9c804721fc6fd1c7e23b84ff0b4bd493f)

Reply:
Test fails here: https://github.com/EspressoSystems/jellyfish/blob/9daf7655bb734ab6c856986363b7389fb1237989/primitives/src/merkle_tree/universal_merkle_tree.rs#L471-L474

Now we may have to think more carefully about the behavior of `universal_forget`. To preserve non-membership proofs, if you forget any leaf from an empty merkle tree, it's not trimming down the memory usage, rather it creates a path to record the information that an empty leaf is forgotten.

Maybe the best option would be allowing user to forget a subtree, but current API doesn't expose the concept of the subtree so it's hard to do so.

Reply:
Yikes. One idea from a long time ago was to have "pinned" paths. Basically an extra boolean on each node, and you are allowed to forget empty sub-trees _unless_ they are pinned. Not entirely sure if that fixes the problem

I think we can save this for future work, and keep the old behavior of forget for now (where it doesn't preserve non-membership proofs). We aren't using forget right now in the sequencer, and with the low memory usage afforded by the persistent updates, I don't know if we'll ever have to. Plus, if any proof is ever missing, we have the ability to fetch it.

Reply:
Rolled back in [6946a32](https://github.com/EspressoSystems/jellyfish/pull/489/commits/6946a320d112ac69e289cc1525bedafaa1219564) and added an issue #495